product_tax_multicompany_default/models/product.py

```diff
@@ -0,0 +1,261 @@
+            field_name_data = "account_sale_tax_id"
+        elif tax_type == "purchase":
+            current_tax = self.supplier_taxes_id
+            field_name_data = "account_purchase_tax_id"
+        propagate_taxes = False
+
+        # Tax-free
+        if not current_tax:
+            # Since all companies can have an empty tax field,if there is data
+            # in the database, it means that we can propagate leaving it empty.
+            if len(dict_product_taxes_bd) > 0:
+                propagate_taxes = True
+        # No tax in other products
+        elif not dict_product_taxes_bd:
+            # If we do not have any taxes in the database and we add a tax, it means
+            # that we can update the product taxes.
+            current_product_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            current_product_taxes_other_companies = set()
+            other_companies = all_companies.filtered(
+                lambda c: c.id != current_company.id
+            )
+            for company in other_companies:
+                current_product_taxes_other_companies.update(
+                    self._taxes_by_company(
+                        field_name_data, company, current_product_tax_ids
+                    )
+                )
+            if len(current_product_taxes_other_companies) > 0:
+                propagate_taxes = True
+        else:
+            current_names = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).mapped("name")
+            current_product_tax_ids = current_tax.filtered(
+                lambda t: t.company_id == current_company
+            ).ids
+            for company_id, value in dict_product_taxes_bd.items():
+                if current_names == list(value.values()):
+                    continue
+                # We are looking to see if there are any taxes that can be applied to
+                # other companies from the taxes that the current product has.
+                current_product_taxes_other_companies = self._taxes_by_company(
+                    field_name_data,
+                    self.env["res.company"].browse(company_id),
+                    current_product_tax_ids,
+                )
+                if len(current_product_taxes_other_companies) == len(current_tax):
+                    propagate_taxes = True
+                    break
+
+        return propagate_taxes
+
+    def _get_product_taxes(self, field):
+        """Returns taxes on purchase or sales taxes
+
+        We need the product taxes for other companies, and we cannot take
+        them from the object itself, so we have to consult the database
+        where the product taxes passed by parameter are located for all
+        companies.
+
+        :param field: can be 'taxes' or 'purchase'
+        """
+        if not self.id:
+            return []
+
+        self.ensure_one()
```

Comment:
If someone pass multiple recordset to the function, i think `self.id` crashes.

```suggestion
        if not self.ids:
            return []

        self.ensure_one()
```